string.c

```diff
@@ -3630,6 +3630,55 @@ rb_str_concat_multi(int argc, VALUE *argv, VALUE str)
     return str;
 }
 
+static VALUE
+rb_str_append_bytes(VALUE str, VALUE str2)
+{
+    Check_Type(str2, T_STRING);
+
+    long str2_len = RSTRING_LEN(str2);
+
+    if (RB_UNLIKELY(str2_len == 0)) {
+        return str;
+    }
+
+    int str_cr = ENC_CODERANGE(str);
+    int str2_cr = ENC_CODERANGE(str2);
+
+    str_buf_cat4(str, RSTRING_PTR(str2), str2_len, true);
+
+    // Fast path for common cases
+    switch (str_cr) {
+      case ENC_CODERANGE_VALID:
+        if (str2_cr == ENC_CODERANGE_VALID) {
+            if (ENCODING_GET_INLINED(str) == ENCODING_GET_INLINED(str2)) {
+                return str;
+            }
+        }
+        else if (str2_cr == ENC_CODERANGE_7BIT && str_enc_fastpath(str)) {
+            return str;
+        }
+        break;
+      case ENC_CODERANGE_7BIT:
+        if (str2_cr == ENC_CODERANGE_7BIT) {
+            return str;
+        }
+        break;
+      case ENC_CODERANGE_UNKNOWN:
+        return str;
+      case ENC_CODERANGE_BROKEN:
+        // rb_str_coderange_scan_restartable isn't capable of repairing a coderange.
+        ENC_CODERANGE_CLEAR(str);
+        return str;
+    }
```

Comment:
Sorry if I'm being ignorant here, but why do we need to have any coderange logic in this function? I thought this function would just preserve/ignore the first operand's encoding?

Reply:
No worries.

Encoding and coderange are two related, but different things. If you append or remove bytes from a string, you MUST update it's coderage. The cheapest way of course it to always reset it to `ENC_CODERANGE_UNKNOWN`, but if you know the coderange of both operands, in many cases you can cheaply compute the resulting coderange without having to scan the string.

We could totally decide to always set it to unknown, it's simple and fast, however if the coderange has to be scanned latter, it will be more costly. So it's a tradeoff.

Also I probably went a bit too far here, maybe it's only work to check for the very common cases and just fallback to either scaning or clearing in other cases.

Reply:
I wrote an alternative version that uses a lookup table for the common and simple/fast cases and fallback to just unknown for others: https://github.com/Shopify/ruby/commit/bfc6eae390fc78c17563f75cac65a8c47214b406